plutus-ledger-api/test/Spec/ScriptSize.hs

```diff
@@ -0,0 +1,203 @@
+    , testCase "V3 Script Size" do
+        let sizeV3 = SBS.length (V3.serialiseCompiledCode codeV3)
+        assertBool "Size V3 script" $ sizeV3 Haskell.> 2000
+    , testCase "V3 Script Size (lazy decoding)" do
+        let sizeV3s = SBS.length (V3.serialiseCompiledCode codeV3lazy)
+        assertBool "Size V3 script with a lazy decoding" $ sizeV3s Haskell.< 100
+    , testCase "V3 script evaluates correctly" do
+        unsafeApplyCode codeV3 (liftCodeDef (V3.toBuiltinData dummyScriptContext))
+          & assertResult unitval
+    , testCase "V3 (lazy) script evaluates correctly" do
+        unsafeApplyCode codeV3lazy (liftCodeDef (V3.toBuiltinData dummyScriptContext))
+          & assertResult unitval
+    ]
+
+codeV2 :: CompiledCode (BuiltinData -> BuiltinData -> BuiltinData -> ())
+codeV2 = $$(compile [||validatorV2||])
+  where
+    validatorV2 :: BuiltinData -> BuiltinData -> BuiltinData -> ()
+    validatorV2 datumBuiltinData redeemerBuiltinData _scriptContext =
+      if expected == redeemer && expected == datum
+        then ()
+        else error ()
+      where
+        redeemer :: Integer
+        redeemer = V2.unsafeFromBuiltinData redeemerBuiltinData
+
+        datum :: Integer
+        datum = V2.unsafeFromBuiltinData datumBuiltinData
+
+codeV3 :: CompiledCode (BuiltinData -> BuiltinUnit)
+codeV3 = $$(compile [||validatorV3||])
+  where
+    validatorV3 :: BuiltinData -> BuiltinUnit
+    validatorV3 scriptContext =
+      if expected == redeemer && Haskell.Just expected == datum
+        then BI.unitval
+        else error ()
+      where
+        redeemer :: Integer
+        redeemer = V3.unsafeFromBuiltinData redeemerBuiltinData
+
+        datum :: Haskell.Maybe Integer
+        datum = V3.unsafeFromBuiltinData . V3.getDatum <$> optionalDatum
+
+        (redeemerBuiltinData, optionalDatum) =
+          case V3.unsafeFromBuiltinData scriptContext of
+            V3.ScriptContext
+              _txInfo
+              (V3.Redeemer redeemerBuiltinData')
+              (V3.SpendingScript _txOutRef optionalDatum') ->
+                (redeemerBuiltinData', optionalDatum')
+            _ -> error ()
+
+codeV3lazy :: CompiledCode (BuiltinData -> BuiltinUnit)
+codeV3lazy = $$(compile [||validatorV3smart||])
+  where
+    validatorV3smart :: BuiltinData -> BuiltinUnit
+    validatorV3smart scriptContext =
+      if expected == redeemer && expected == datum
+        then BI.unitval
+        else error ()
+      where
+        redeemerFollowedByScriptInfo :: BI.BuiltinList BuiltinData
+        redeemerFollowedByScriptInfo = BI.tail (constrArgs scriptContext)
+
+        redeemerBuiltinData :: BuiltinData
+        redeemerBuiltinData = BI.head redeemerFollowedByScriptInfo
+
+        scriptInfoData :: BuiltinData
+        scriptInfoData = BI.head (BI.tail redeemerFollowedByScriptInfo)
+
+        datumData :: BuiltinData
+        datumData = BI.head (constrArgs (BI.head (BI.tail (constrArgs scriptInfoData))))
```

Comment:
That's so ugly and tedious. We should probably provide a dedicated data type for doing this lazy decoding (`data JustGiveMeDatumAndRedeemer = <...>`), so that `unsafeFromBuiltinData` does the right thing for it.